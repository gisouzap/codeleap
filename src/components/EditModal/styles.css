.modal-wrapper,
.input-group {
  display: flex;
  flex-direction: column;
}

.modal-wrapper {
  gap: 24px;
}

.input-group {
  gap: 8px;
}
