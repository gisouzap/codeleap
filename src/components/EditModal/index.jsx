import { useCallback, useEffect } from 'react';
import PropTypes from 'prop-types';

import Typography from '../Typography';
import TextField from '../TextField';

import Button from '../Button';
import Card from '../Card';

import './styles.css';

function EditModal({ post, onChange, onSave, onClose }) {
  const handleEscape = useCallback(
    event => {
      if (event.keyCode === 27) {
        onClose();
      }
    },
    [onClose]
  );

  useEffect(() => {
    document.addEventListener('keydown', handleEscape, false);
    return () => {
      document.removeEventListener('keydown', handleEscape, false);
    };
  }, [handleEscape]);

  if (!post) {
    return null;
  }

  return (
    <div className="modal">
      <Card style={{ width: '670px' }}>
        <div className="modal-wrapper">
          <Typography variant="title">Edit item</Typography>

          <div className="input-group">
            <Typography variant="label">Title</Typography>
            <TextField
              onChange={e => onChange({ title: e.target.value })}
              value={post.title}
            />
          </div>
          <div className="input-group">
            <Typography variant="label">Content</Typography>
            <textarea
              value={post.content}
              onChange={e => onChange({ content: e.target.value })}
              className="textarea"
            />
          </div>
        </div>
        <div className="buttons">
          <Button variant="light" onClick={onClose} label="Cancel" />
          <Button variant="success" onClick={onSave} label="Save" />
        </div>
      </Card>
    </div>
  );
}

export default EditModal;

EditModal.propTypes = {
  onSave: PropTypes.func.isRequired,
  onClose: PropTypes.func.isRequired,
  onChange: PropTypes.func.isRequired,
  post: PropTypes.shape({
    title: PropTypes.string,
    content: PropTypes.string,
  }),
};
