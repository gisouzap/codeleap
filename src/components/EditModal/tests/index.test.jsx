import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import EditModal from '../index';

describe('EditModal', () => {
  const handleSave = jest.fn();
  const handleChange = jest.fn();
  const handleClose = jest.fn();

  const post = {
    title: 'CodeLeap',
    content: 'Hello world',
  };

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should not log errors in console', () => {
    const spy = jest.spyOn(global.console, 'error');

    render(
      <EditModal
        post={post}
        onChange={handleChange}
        onSave={handleSave}
        onClose={handleClose}
      />
    );
    expect(spy).not.toHaveBeenCalled();
  });

  it('should call handleClose when ESC key is pressed', () => {
    const { container } = render(
      <EditModal
        post={post}
        onChange={handleChange}
        onSave={handleSave}
        onClose={handleClose}
      />
    );

    fireEvent.keyDown(container.querySelector('.modal'), { keyCode: 27 });

    expect(handleClose).toHaveBeenCalled();
  });

  it('should not call handleClose when other key is pressed', () => {
    const { container } = render(
      <EditModal
        post={post}
        onChange={handleChange}
        onSave={handleSave}
        onClose={handleClose}
      />
    );

    fireEvent.keyDown(container.querySelector('.modal'), { keyCode: 25 });

    expect(handleClose).not.toHaveBeenCalled();
  });

  it('should call handleChange with new title', () => {
    const { getAllByRole } = render(
      <EditModal
        post={post}
        onChange={handleChange}
        onSave={handleSave}
        onClose={handleClose}
      />
    );

    const input = getAllByRole('textbox')[0];

    fireEvent.change(input, { target: { value: 'New value' } });

    expect(handleChange).toHaveBeenCalledWith({ title: 'New value' });
  });

  it('should call handleChange with new content', () => {
    const { getAllByRole } = render(
      <EditModal
        post={post}
        onChange={handleChange}
        onSave={handleSave}
        onClose={handleClose}
      />
    );

    const textarea = getAllByRole('textbox')[1];

    fireEvent.change(textarea, { target: { value: 'New content' } });

    expect(handleChange).toHaveBeenCalledWith({ content: 'New content' });
  });

  it('should call handleSave when save button is clicked', () => {
    render(
      <EditModal
        post={post}
        onChange={handleChange}
        onSave={handleSave}
        onClose={handleClose}
      />
    );

    const saveBtn = screen.getByRole('button', { name: /save/i });

    fireEvent.click(saveBtn);

    expect(handleSave).toHaveBeenCalledTimes(1);
  });

  it('should render and match the snapshot equal null', () => {
    const { container } = render(
      <EditModal
        onChange={handleChange}
        onSave={handleSave}
        onClose={handleClose}
      />
    );

    expect(container).toMatchSnapshot();
  });

  it('should render and match the snapshot', () => {
    const { container } = render(
      <EditModal
        post={post}
        onChange={handleChange}
        onSave={handleSave}
        onClose={handleClose}
      />
    );

    expect(container).toMatchSnapshot();
  });
});
