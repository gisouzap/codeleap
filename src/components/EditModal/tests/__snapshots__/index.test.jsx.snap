// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditModal should render and match the snapshot 1`] = `
<div>
  <div
    class="modal"
  >
    <div
      class="card"
      style="width: 670px;"
    >
      <div
        class="modal-wrapper"
      >
        <p
          class="title"
        >
          Edit item
        </p>
        <div
          class="input-group"
        >
          <p
            class="label"
          >
            Title
          </p>
          <input
            class="input"
            type="text"
            value="CodeLeap"
          />
        </div>
        <div
          class="input-group"
        >
          <p
            class="label"
          >
            Content
          </p>
          <textarea
            class="textarea"
          >
            Hello world
          </textarea>
        </div>
      </div>
      <div
        class="buttons"
      >
        <button
          class="button light"
          type="button"
        >
          Cancel
        </button>
        <button
          class="button success"
          type="button"
        >
          Save
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`EditModal should render and match the snapshot equal null 1`] = `<div />`;
